File: simplify_path.js
#!/usr/bin/env node

function simplify_iterative(path) {
  var stack = [];
  var grounded = false;
  if (path[0] === '/') {
    stack.push('');
    grounded = true;
  }
  var dirs = path.split('/');
  while (dirs.length > 0) {
    var dir = dirs.shift();
    if (dir === '.' || dir == '') {
      // no-op
    } else if (dir === '..') {
      if (grounded === true) {
        stack.pop();
      } else {
        stack.push(dir);
      }
    } else {
      grounded = true;
      stack.push(dir);
    }
  }
  if (path[path.length - 1] === '/') {
    stack.push('');
  }
  return stack.join('/');
}

var path_1 = '/usr/lib/../bin/gcc';
var path_2 = 'sctipts//./../scripts/awkscripts/././';
var path_3 = '../../export/workspace/../goodies/';

console.log(simplify_iterative(path_1));
console.log(simplify_iterative(path_2));
console.log(simplify_iterative(path_3));

/*

/usr/bin/gcc
scripts/awkscripts/
../../export/goodies/

*/

File: simplify_path.rb
#!/usr/bin/env ruby

def simplify_iterative(path)
  stack = []
  grounded = false
  if path[0] == '/'
    stack.push('')
    grounded = true
  end
  dirs = path.split('/')
  while dirs.length > 0
    dir = dirs.shift
    if dir == '.' or dir == ''
      # no-op
    elsif dir == '..'
      if (grounded == true)
        stack.pop()
      else
        stack.push(dir)
      end
    else
      grounded = true
      stack.push(dir)
    end
  end
  if path[-1] == '/'
    stack.push('')
  end
  return stack.join('/')
end

path_1 = '/usr/lib/../bin/gcc'
path_2 = 'sctipts//./../scripts/awkscripts/././'
path_3 = '../../export/workspace/../goodies/'

puts simplify_iterative path_1
puts simplify_iterative path_2
puts simplify_iterative path_3

__END__

/usr/bin/gcc
scripts/awkscripts/
../../export/goodies/
